packages/legend-query-builder/src/stores/filter/QueryBuilderFilterState.ts

```diff
@@ -164,90 +320,86 @@ export class FilterConditionState implements Hashable {
-      `Can't find an operator for property '${this.propertyExpressionState.path}': no operators registered`,
-    );
-    if (!newCompatibleOperators.includes(this.operator)) {
-      this.changeOperator(
-        newCompatibleOperators[0] as QueryBuilderFilterOperator,
-      );
-    } else if (!this.operator.isCompatibleWithFilterConditionValue(this)) {
-      this.setValue(this.operator.getDefaultFilterConditionValue(this));
-    }
-  }
-
   changeOperator(val: QueryBuilderFilterOperator): void {
     this.setOperator(val);
     if (!this.operator.isCompatibleWithFilterConditionValue(this)) {
       let defaultValue = this.operator.getDefaultFilterConditionValue(this);
+      // Don't allow invalid InstanceValues or empty strings to be set as list element
       if (
         defaultValue instanceof CollectionInstanceValue &&
-        this.value instanceof InstanceValue &&
-        isValidInstanceValue(this.value)
+        this.rightConditionValue instanceof
+          FilterValueSpecConditionValueState &&
+        this.rightConditionValue.value instanceof InstanceValue &&
+        isValidInstanceValue(this.rightConditionValue.value) &&
+        this.rightConditionValue.value.values[0] !== ''
       ) {
         instanceValue_setValues(
           defaultValue,
-          [this.value],
+          [this.rightConditionValue.value],
           this.filterState.queryBuilderState.observerContext,
         );
       } else if (
         defaultValue instanceof InstanceValue &&
-        this.value instanceof CollectionInstanceValue &&
-        this.value.values.length
+        this.rightConditionValue instanceof
+          FilterValueSpecConditionValueState &&
+        this.rightConditionValue.value instanceof CollectionInstanceValue &&
+        this.rightConditionValue.value.values.length
       ) {
-        defaultValue = this.value.values[0];
+        defaultValue = this.rightConditionValue.value.values[0];
       }
-      this.setValue(defaultValue);
+      this.buildFromValueSpec(defaultValue);
     }
   }
 
   setOperator(val: QueryBuilderFilterOperator): void {
     this.operator = val;
   }
 
-  setValue(val: ValueSpecification | undefined): void {
-    this.value = val
-      ? observe_ValueSpecification(
-          val,
-          this.filterState.queryBuilderState.observerContext,
-        )
-      : undefined;
+  setRightConditionValue(val: FilterConditionValueState | undefined): void {
+    this.rightConditionValue = val;
   }
 
   addExistsLambdaParamNames(val: string): void {
     this.existsLambdaParamNames.push(val);
   }
 
+  buildFromValueSpec(val: ValueSpecification | undefined): void {
+    if (
+      this.rightConditionValue instanceof FilterValueSpecConditionValueState
+    ) {
+      this.rightConditionValue.setValue(val);
+    } else {
+      this.setRightConditionValue(
+        new FilterValueSpecConditionValueState(this, val),
+      );
+    }
+  }
+
+  buildFromPropertyExpressionState(
```

Comment:
do you think if it's better if we could rename this to `buildRightConditionValueFromPropertyExpressionState` ?

Reply:
Done!